examples/vision/focal_modulation_network.py

```diff
@@ -0,0 +1,1220 @@
+constraints interactions in a local region with a shared kernel.
+
+The **attention** module is mathematically written as
+
+$$\text{Attention} = \text{softmax}(\frac{QK^{T}}{\sqrt{d_k}})V$$
+
+where:
+- $Q$ is the query
+- $K$ is the key
+- $V$ is the value
+- $d_k$ is the dimension of the key
+
+With **self-attention** the query, key, and value are all sourced from the input
+sequence. Let's rewrite the attention equation for self-attention.
+
+$$\text{Self Attention} = \text{softmax}(\frac{QQ^{T}}{\sqrt{d_q}})Q$$
+
+Upon looking at the equation of self-attention we see that it is a quadratic equation.
+This means that as number of tokens increase, so does the computation time (cost too). To
+mitigate this problem, and also to make the Transformer more interpretable Yang et. al
+have tried to replace the Self Attention module with better componenets.
+
+## The Solution
+
+Yang et. al introduce the revolutionary Focal Modulation _Layer_, poised to serve as a
+seamless replacement for the Self Attention Layer. The layer boasts high
+interpretability, making it a valuable tool for Deep Learning practitioners.
+
+In this tutorial, we will delve into the practical application of this layer by training
+the entire model on the CIFAR-10 dataset and visually interpreting the layer's
+performance.
+
+Note: We have tried aligning our implementation with the
+[official implementation](https://github.com/microsoft/FocalNet) as
+much as we could.
+"""
+
+"""
+# Setup and Imports
+
+`tensorflow-addons` is required as we will use the AdamW optimizer to train our model.
+"""
+
+"""shell
+!pip install --upgrade --quiet tensorflow-addons
+"""
+
+import numpy as np
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+from tensorflow_addons.optimizers import AdamW
+
+from typing import Optional, Tuple, List
+from matplotlib import pyplot as plt
+from random import randint
+
+# Set seed for reproducibility.
+tf.keras.utils.set_random_seed(42)
+
+"""
+# Global Configuration
+
+We do not have any strong ratioanle behind choosing these hyperparameters. Please feel
+free to change the configuration and train the model.
+"""
+
+# DATA
+BUFFER_SIZE = 2048
+BATCH_SIZE = 1024
+AUTO = tf.data.AUTOTUNE
+INPUT_SHAPE = (32, 32, 3)
+IMAGE_SIZE = 48
+NUM_CLASSES = 10
+
+# OPTIMIZER
+LEARNING_RATE = 1e-4
+WEIGHT_DECAY = 1e-4
+
+# TRAINING
+EPOCHS = 50
+
+"""
+# Load and process the CIFAR-10 dataset
+"""
+
+(x_train, y_train), (x_test, y_test) = keras.datasets.cifar10.load_data()
+(x_train, y_train), (x_val, y_val) = (
+    (x_train[:40000], y_train[:40000]),
+    (x_train[40000:], y_train[40000:]),
+)
+print(f"Training samples: {len(x_train)}")
+print(f"Validation samples: {len(x_val)}")
+print(f"Testing samples: {len(x_test)}")
+
+"""
+## Build the augmentations
+
+We have used the `keras.Sequential` API to compose all the individual augmentation steps
```

Comment:
In general, always use present tense, e.g. `we use` and not `we have used` nor `we will use`